utility/SPIDEV/gpio.cpp

```diff
@@ -27,152 +32,83 @@ GPIO::~GPIO()
 {
 }
 
-void GPIO::open(int port, int DDR)
-{
-    FILE* f;
-    f = fopen("/sys/class/gpio/export", "w");
-    if (f == NULL) {
-        throw GPIOException("can't export GPIO pin .check access rights");
-    }
-    fprintf(f, "%d\n", port);
-    fclose(f);
-
-    int counter = 0;
-    char file[128];
-    sprintf(file, "/sys/class/gpio/gpio%d/direction", port);
-
-    while ((f = fopen(file, "w")) == NULL) { //Wait 10 seconds for the file to be accessible if not open on first attempt
-        sleep(1);
-        counter++;
-        if (counter > 10) {
-            throw GPIOException("can't access /sys/class/gpio/gpio%d/direction GPIO pin. check access rights");
-            /*perror("Could not open /sys/class/gpio/gpio%d/direction");
-            exit(0); */
-        }
-    }
-    int l = (DDR == 0) ? fprintf(f, "in\n") : fprintf(f, "out\n");
-    if (!(l == 3 || l == 4)) {
-        fclose(f);
-        throw GPIOException("can't set direction on GPIO pin. check access rights");
-    }
-    /*
-    if (DDR == 0)
-        fprintf(f, "in\n");
-    else
-        printf(f, "out\n");
-    */
-    fclose(f);
+int fd;
 
-    // Caches the GPIO descriptor;
-    sprintf(file, "/sys/class/gpio/gpio%d/value", port);
-    int flags = (DDR == 0) ? O_RDONLY : O_WRONLY;
-    int fd = ::open(file, flags);
-    if (fd < 0) {
-        throw GPIOException("Can't open the GPIO");
+void GPIO::Gopen(int port, int DDR)
+{
+    fd = open(dev_name, O_RDONLY);
+    if (fd >= 0) {
+        close(fd);
     }
     else {
-        cache[port] = fd; // cache the fd;
-        lseek(fd, SEEK_SET, 0);
+        dev_name = "/dev/gpiochip0";
+        fd = open(dev_name, O_RDONLY);
+        if (fd >= 0) {
+            close(fd);
+        }
+        else {
+            throw GPIOException("can't open /dev/gpiochip");
+        }
     }
 }
 
-void GPIO::close(int port)
+void GPIO::Gclose(int port)
 {
-    std::map<int, GPIOfdCache_t>::iterator i;
-    i = cache.find(port);
-    if (i != cache.end()) {
-        close(i->second); // close the cached fd
-        cache.erase(i);   // Delete cache entry
-    }
-    // Do unexport
-    FILE* f;
-    f = fopen("/sys/class/gpio/unexport", "w");
-    if (f != NULL) {
-        fprintf(f, "%d\n", port);
-        fclose(f);
-    }
 }
```

Comment:
Its odd that this is now a non-op. I don't know if we should instead return the state of the specified pin to a default (or on-boot) state.

Reply:
Of course, we aren't exposing this via Arduino-like macros, so maybe it doesn't matter.

Reply:
I don't know. The GPIO file descriptors are all closed in the read/write functions so it probably doesn't matter. 
I thought of trying to store the FD and info and just toggle the GPIOs when necessary if possible, in which case it would be needed, but with the way it works now I don't think it is.

Reply:
Maybe we can reduce any possible unnecessary power consumption by setting the specified pin to an output in an inactive state (LOW).

There might be some discrepancies about the pin's default active state. For example, the SPI CS pins should have a ACTIVE_LOW state as opposed to the normal ACTIVE_HIGH for actual general purpose pins.

I still need to research this idea, but I doubt there's a problem with just calling
`GPIO::write(port, OUTPUT_LOW);` here

Reply:
I am overthinking this. Most RF24 usage wouldn't ever call `GPIO::close()` explicitly. My idea for calling `GPIO::write(port, OUTPUT_LOW)` would imply a destructor in RF24 class to de-init pins used on Linux.

We can revisit this if/when there's a problem. Resolving this.